lib/useQuery.js

```diff
@@ -31,252 +38,138 @@ export function useQuery(serviceName, options = {}, queryHookOptions = {}) {
     throw new Error(`Bad realtime option, must be one of ${[realtimeModes].join(', ')}`)
   }
 
   if (!fetchPolicies.includes(fetchPolicy)) {
     throw new Error(`Bad fetchPolicy option, must be one of ${[fetchPolicies].join(', ')}`)
   }
 
-  const queryId = `${method.substr(0, 1)}:${hashObject({
+  const queryId = useQueryHash({
     serviceName,
     method,
     id,
     params,
+    allPages,
     realtime,
-  })}`
+  })
 
-  let [cachedResult, updateCache] = useCache({
-    serviceName,
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  params = useMemo(() => params, [queryId])
+
+  const [cachedResult, updateCache] = useCache({
     queryId,
+    serviceName,
     method,
-    id,
     params,
     realtime,
     selectData,
-    transformResponse,
     matcher,
   })
 
-  let hasCachedData = !!cachedResult.data
-  const fetched = fetchPolicy === 'cache-first' && hasCachedData
+  const isCacheSufficient = fetchPolicy === 'cache-first' && !!cachedResult
 
   const [state, dispatch] = useReducer(reducer, {
-    reloading: false,
-    fetched,
-    fetchedCount: 0,
-    refetchSeq: 0,
+    status: isCacheSufficient ? 'success' : 'pending',
     error: null,
   })
 
-  if (fetchPolicy === 'network-only' && state.fetchedCount === 0) {
-    cachedResult = emptyCachedResult
-    hasCachedData = false
-  }
+  const isPending = state.status === 'pending'
 
-  const handleRealtimeEvent = useCallback(
-    payload => {
-      if (realtime !== 'refetch') return
-      dispatch({ type: 'refetch' })
-    },
-    [dispatch, realtime],
-  )
+  const requestRef = useRef(0)
 
   useEffect(() => {
-    let disposed = false
-
-    if (state.fetched) return
     if (skip) return
+    if (!isPending) return
+    if (isCacheSufficient) return
 
-    dispatch({ type: 'fetching' })
-    const service = feathers.service(serviceName)
-    const result =
-      method === 'get'
-        ? get(service, id, params, { queryId })
-        : find(service, params, { queryId, allPages, parallel })
+    // increment the request ref so we can ignore old requests
+    const reqRef = (requestRef.current = requestRef.current + 1)
 
-    result
+    dispatch({ type: 'fetching' })
+    fetch(feathers, serviceName, method, id, params, { queryId, allPages, parallel })
       .then(res => {
-        // no res means we've piggy backed on an in flight request
-        if (res) {
-          updateCache(res)
-        }
-
-        if (!disposed) {
-          dispatch({ type: 'success' })
-        }
+        flushSync(() => {
+          // no res means we've piggy backed on an in flight request
+          if (res) {
+            // update cache even if this particular useQuery invocation
+            // no longer needs the result, that is because we are potentially
+            // sharing this request with other useQuery invocations that
+            // might still want the result to be propagated to cache
+            updateCache(transformResponse(res))
```

Comment:
I think we should also ignore this response if this is for an old request with different query?

Reply:
I considered that, and even made the code do it, but then remembered why it is this way - it's detailed in the comment. Because there's the inflight mechanic, you might call useFind() in 5 different places at once, only one of them fetches+gets response passed in+calls updateCache, the others _rely_ on that one to update the cache. So if that 1 is cancelled, but the other 4 are still mounted, they would end up hanging/crashing with success state but no cached result.

Now there is probably a way to make that a bit "cleaner" / decoupled from the request maybe, but couldn't see a quick way to do that.

Does that make sense?

Reply:
Aah yep got it. Maybe `inflight` can always resolve with the result regardless of whether it was the first request, and then the cache can ignore updates its already seen (which it could do by checking the reference). Or maybe `fetch` should handle updating the cache

But this looks fine

Reply:
Inflight now always resolves result, but cache ignores repeated updates. This is neat cause if you cancel a request, we don't update cache and also further simplified the useQuery implementation.

Reply:
nice!